crates/pico_macros/src/memo.rs

```diff
@@ -0,0 +1,134 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, FnArg, ItemFn, PatType, ReturnType};
+
+pub(crate) fn memo(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        sig,
+        vis,
+        block,
+        attrs,
+    } = parse_macro_input!(item as ItemFn);
+
+    let fn_name = sig.ident.clone().to_string();
+
+    let inputs = sig.inputs.iter().collect::<Vec<_>>();
+
+    if inputs.is_empty() {
+        return syn::Error::new_spanned(&sig, "Function must have at least one argument")
+            .to_compile_error()
+            .into();
+    }
+
+    let db_arg = match &inputs[0] {
+        FnArg::Typed(PatType { pat, .. }) => pat.clone(),
+        _ => {
+            return syn::Error::new_spanned(inputs[0], "Expected a typed argument")
+                .to_compile_error()
+                .into()
+        }
+    };
+
+    let other_args = inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { pat, .. }) => pat,
+        _ => unreachable!(),
+    });
+
+    let argument_types = inputs.iter().skip(1).map(|arg| match arg {
```

Comment:
nit: if we make this mutable, we can avoid a clone below. Likewise for `let other_args`